### PR TITLE
feat: allow admins to enable the usage of lua standard library (#6050)

### LINKED ISSUE #6050: Allow administrator enable lua standard library for health checks.
# Summary

For security reasons standard library is disabled lua based health check. However, this is blocking the implementation of complex health checks that requires functions like `string.find` or `string.match`. (https://github.com/argoproj/argo-cd/issues/6039) . 

# Motivation

The argocd administrator should be able to enable the standard library after reviewing the script.

# Proposal

* Enable standard library for built-in health checks
* Introduce `health.lua.useOpenLibs: true` setting that could be set be Argo CD admin and enable standard library  support for a specific health check:
```yaml
apiVersion: v1
data:
  resource.customizations: |
    kubernetes-client.io/ExternalSecret:
      health.lua.useOpenLibs: true
      health.lua: | ....
```

### PR DESCRIPTION
**Pull request by chetan-rns**
Currently, the usage of the standard Lua library is always disabled, making it difficult to implement complex health check scripts.
This feat allows admins to control the usage of the standard library by setting `health.lua.useOpenLibs` (merged-keys)/ `resource.customizations.useOpenLibs.<group_kind>`(split keys) in argocd-cm ConfigMap.

Example:

Merged keys:

```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: argocd-cm
data:
  resource.customizations: |
    ServiceAccount:
      health.lua.useOpenLibs: true
      health.lua: |
      .............
```

Split keys:

```yaml
data:
  resource.customizations.useOpenLibs.ServiceAccount: "true"
  resource.customizations.health.ServiceAccount:
    -- Lua standard libraries are enabled for this script
```

In the above example, usage of standard libs will be allowed for the ServiceAccount health script.

Closes: #6050 

Signed-off-by: Chetan Banavikalmutt <chetanrns1997@gmail.com>

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [x] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by chetan-rns** [comment]
Hi @alexmt @jannfis, PTAL :)

File: util/lua/lua.go [hunk 1]
```diff
@@ -39,6 +39,8 @@ type ResourceHealthOverrides map[string]appv1.ResourceOverride
 func (overrides ResourceHealthOverrides) GetResourceHealth(obj *unstructured.Unstructured) (*health.HealthStatus, error) {
 	luaVM := VM{
 		ResourceOverrides: overrides,
+		// check if the usage of lua standard library is enabled
+		UseOpenLibs: overrides.isOpenLibsEnabled(obj),
 	}
 	script, err := luaVM.GetHealthScript(obj)
```
Comment (alexmt) [inline util/lua/lua.go:45]: The `UseOpenLibs` should be `true` for all built-in scripts. 

I think you might return `isBuiltIn` boolean from `GetHealthScript` and use it to enable/disable `UseOpenLibs`

https://github.com/argoproj/argo-cd/blob/45713b1ffca5944143a68435dff43eda071286ae/util/lua/lua.go#L130

**Review by alexmt — CHANGES_REQUESTED** [review]
Thank you @chetan-rns !

Added one minor improvement.

File: util/lua/lua.go (hunk 1 above)
Comment (chetan-rns) [reply util/lua/lua.go]: Thanks! Made the changes

**Review by jannfis — CHANGES_REQUESTED** [review]
Can you please bring the changes from this PR in-line with changes introduced in #6060

**Comment by chetan-rns** [comment]
@alexmt @jannfis  Updated the PR to accommodate the new split-key conventions, the cm key for enabling open libs looks like `resource.customizations.useOpenLibs.<group_kind>`.  PTAL

Ex: 
```yaml
data:
  resource.customizations.useOpenLibs.apps_Deployment: "false"  # only allows libs for Deployment resource
```

**Review by alexmt — APPROVED** [review]
Thank you ! LGTM

@jannfis , is it good to merge?

**Review by jannfis — APPROVED** [review]
LGTM, thank you @chetan-rns !

**Pull request merged**